server/src/main/java/org/elasticsearch/index/engine/SegmentsStats.java

```diff
@@ -48,14 +51,7 @@ public SegmentsStats(StreamInput in) throws IOException {
         versionMapMemoryInBytes = in.readLong();
         bitsetMemoryInBytes = in.readLong();
         maxUnsafeAutoIdTimestamp = in.readLong();
-
-        final int size = in.readVInt();
-        final ImmutableOpenMap.Builder<String, FileStats> files = ImmutableOpenMap.builder(size);
-        for (int i = 0; i < size; i++) {
-            FileStats file = new FileStats(in);
-            files.put(file.getExt(), file);
-        }
-        this.files = files.build();
+        files = in.readList(FileStats::new).stream().collect(Collectors.toMap(FileStats::getExt, Function.identity()));
```

Comment:
This one could be a little hot in some cases I believe. Let's keep the previous pattern and pre-size with `org.elasticsearch.common.util.Maps#newMapWithExpectedSize` that does without so many allocations for the list/stream etc.?